java/integTest/src/test/java/glide/standalone/PubSubTests.java

```diff
@@ -0,0 +1,827 @@
+
+// @Timeout(30) // sec
+public class PubSubTests {
+
+    // TODO protocol version
+    @SneakyThrows
+    @SuppressWarnings("unchecked")
+    private <M extends ChannelMode> BaseClient createClientWithSubscriptions(
+            boolean standalone,
+            Map<M, Set<String>> subscriptions,
+            Optional<MessageCallback> callback,
+            Optional<Object> context) {
+        if (standalone) {
+            var subConfigBuilder =
+                    StandaloneSubscriptionConfiguration.builder()
+                            .subscriptions((Map<PubSubChannelMode, Set<String>>) subscriptions);
+
+            if (callback.isPresent()) {
+                subConfigBuilder.callback(callback.get(), context.get());
+            }
+            return RedisClient.CreateClient(
+                            commonClientConfig().subscriptionConfiguration(subConfigBuilder.build()).build())
+                    .get();
+        } else {
+            var subConfigBuilder =
+                    ClusterSubscriptionConfiguration.builder()
+                            .subscriptions((Map<PubSubClusterChannelMode, Set<String>>) subscriptions);
+
+            if (callback.isPresent()) {
+                subConfigBuilder.callback(callback.get(), context.get());
+            }
+
+            return RedisClusterClient.CreateClient(
+                            commonClusterClientConfig()
+                                    .subscriptionConfiguration(subConfigBuilder.build())
+                                    .build())
+                    .get();
+        }
+    }
+
+    private <M extends ChannelMode> BaseClient createClientWithSubscriptions(
+            boolean standalone, Map<M, Set<String>> subscriptions) {
+        return createClientWithSubscriptions(
+                standalone, subscriptions, Optional.empty(), Optional.empty());
+    }
+
+    @SneakyThrows
+    private BaseClient createClient(boolean standalone) {
+        if (standalone) {
+            return RedisClient.CreateClient(commonClientConfig().build()).get();
+        }
+        return RedisClusterClient.CreateClient(commonClusterClientConfig().build()).get();
+    }
+
+    /** Message queue used in callback to analyze received messages. Number is a client ID. */
+    private final ConcurrentLinkedDeque<Pair<Integer, Message>> mq = new ConcurrentLinkedDeque<>();
+
+    /** Clients used in a test. */
+    private final List<BaseClient> clients = new ArrayList<>();
+
+    @BeforeEach
+    @SneakyThrows
+    public void cleanup() {
+        mq.clear();
+        for (var client : clients) {
+            if (client instanceof RedisClusterClient) {
+                ((RedisClusterClient) client).customCommand(new String[] {"unsubscribe"}).get();
+                ((RedisClusterClient) client).customCommand(new String[] {"punsubscribe"}).get();
+                ((RedisClusterClient) client).customCommand(new String[] {"sunsubscribe"}).get();
```

Comment:
is this necessary, or do we clean it up when we destroy the client?

Reply:
Maybe it is not needed, but I want to explicitly unsubscribe in case if a client isn't GC'ed and remains active/connected.

Reply:
The client gets close()'d  right after these commands run, so I wouldn't think there's a risk of it not getting cleaned up here (not necessarily GC'd).